File: Java/t101.java
package app;

/**
 * Author:lyc
 */

/**
 * Definition for a binary tree node.
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) { this.val = val; }
}

class Solution {
    public boolean isSymmetric(TreeNode root) {

        if (root == null){
            return true;
        }else {
            return checkSame(root.left, root.right);
        }
    }

    private boolean checkSame(TreeNode left, TreeNode right){

        if(left != null && right != null && left.val == right.val){
            return checkSame(left.left, right.right) && checkSame(left.right, right.left);
        }
        else if(left == null && right == null){
            return true;
        }
        else {
            return false;
        }
    }
}

public class t101 {

    public static void main(String[] args) {

        TreeNode p = new TreeNode(1);
        p.left = new TreeNode(2);
        p.right = new TreeNode(2);
        p.left.left = new TreeNode(3);
        p.right.right = new TreeNode(3);
        p.left.right = new TreeNode(4);
        p.right.left = new TreeNode(4);

        Solution result = new Solution();
        boolean res = result.isSymmetric(p);
        System.out.println(res);

    }
}

File: Python/t101.py
# -*- conding:utf-8 -*-
#Author:lyc
import os, sys

# 题目：
# 给定一个二叉树，检查它是否是镜像对称的。

# 示例：
# 输入：[1,2,2,3,4,4,3]
# 输出: true
# 输入：[1,2,2,null,3,null,3]
# 输出: false

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if root == None:
            return True
        else:
            return self.checkSame(root.left, root.right)

    def checkSame(self, left:TreeNode, right:TreeNode):

        if left != None and right != None and left.val == right.val:
            return self.checkSame(left.right, right.left) and \
                   self.checkSame(left.left, right.right)
        elif left == None and right == None:
            return True
        else:
            return False


p = TreeNode(1)
p.left = TreeNode(2)
p.right = TreeNode(2)
p.left.left = TreeNode(3)
p.left.right = TreeNode(4)
p.right.left = TreeNode(4)
p.right.right = TreeNode(3)


result = Solution()
print(result.isSymmetric(p))
